### DISCUSSION TITLE
How can I understand if client token is expired?

### DISCUSSION THREAD

**Discussion by lego963** [gophercloud/gophercloud#2423] [Q&A] [answered]
I created the Identity Service client and I cached it.
Can I somehow check if token in the client is still valid?
```go
	pClient, err := openstack.AuthenticatedClient(opts)
	if err != nil {
		return fmt.Errorf("error creating provider client: %w", err)
	}

	sClient, err := openstack.NewIdentityV3(pClient, gophercloud.EndpointOpts{})
	if err != nil {
		return fmt.Errorf("error creating service client: %w", err)
	}
```
Can I get `expiresAt` attr?

**Comment by kayrus** [comment, marked as answer]
You can get `AuthResult` from `p.Client.Provider.GetAuthResult()` and then assert it to `res.(*v3/tokens.GetResults`, afterwards you can directly access https://pkg.go.dev/github.com/gophercloud/gophercloud@v0.25.0/openstack/identity/v3/tokens#GetResult.ExtractToken
https://pkg.go.dev/github.com/gophercloud/gophercloud@v0.25.0/openstack/identity/v3/tokens#Token
FYI: https://github.com/gophercloud/gophercloud/issues/2111

**Reply by lego963** [reply]
thanks for support, I did almost the same
```go
	tokenResponse := tokens.Get(sClient, sClient.Token())
	token, err := tokenResponse.ExtractToken()
	if err != nil {
		return err
	}
```
And here I can get `expiresAt` :)
